### PR TITLE
bpo-32075: Expose ZipImporter Type Object in the include header files.

### PR DESCRIPTION
**Pull request by AraHaan**
This should hopefully allow subclassing the zipimporter class outside of
the Python layer as well.

See the issue at bpo for more information as to why this needs exposed in the include files.

https://bugs.python.org/issue32075

### CODE REVIEW

File: Include/zipimport.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+struct _zipimporter {
+    PyObject_HEAD
+    PyObject *archive;  /* pathname of the Zip archive,
+                           decoded from the filesystem encoding */
+    PyObject *prefix;   /* file prefix: "a/sub/directory/",
+                           encoded to the filesystem encoding */
+    PyObject *files;    /* dict with file info {path: toc_entry} */
+};
+
+typedef struct _zipimporter ZipImporter;
```
Comment (AraHaan) [inline Include/zipimport.h]: It seems that somehow when including this file to test this that the following error sometimes happens:

```
zipimport.h(19): error C2079: 'ZipImporter_Type' uses undefined struct '_typeobject'
```

File: Modules/clinic/zipimport.c.h [hunk 2]
```diff
@@ -32,7 +32,7 @@ zipimport_zipimporter___init__(PyObject *self, PyObject *args, PyObject *kwargs)
     int return_value = -1;
     PyObject *path;
 
-    if ((Py_TYPE(self) == &ZipImporter_Type) &&
+    if ((Py_TYPE(self) == &PyZipImporter_Type) &&
```
Comment (AraHaan) [inline Modules/clinic/zipimport.c.h]: @serhiy-storchaka I think I am having trouble updating the clinic stuff on this so it dont revert the changes to this file and build on travis properly.

File: Include/zipimport.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#ifndef Py_ZIPIMPORT_H
+#define Py_ZIPIMPORT_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+PyAPI_DATA(PyTypeObject) PyZipImporter_Type;
```
Comment (brettcannon) [inline Include/zipimport.h:7]: This and the rest of the file shouldn't be part of the stable ABI.

**Comment by brettcannon** [comment]
I personally don't have time to review and contemplate the ramifications of this, but I would suggest that any changes not be part of the stable ABI.

File: Include/zipimport.h (hunk 3 above)
Comment (AraHaan) [reply Include/zipimport.h:7]: But then how would someone be able to subclass or subtype the zip importer in their C code?

Comment (brettcannon) [reply Include/zipimport.h:7]: Relying only on the stable ABI is an opt-in thing when you build an extension, so leaving this out of it won't prevent it from being available overall, just in a certain instance.
